mne/stats/cluster_level.py

```diff
@@ -1880,23 +1886,18 @@ def cluster_test(
             raise ValueError("for paired t-test, each subject must have 2 observations")
 
     # extract the data from the dataframe
-    def _extract_data_array(series):
+    outer_func = np.concatenate if is_epo else np.array
+    axes = (-3, -1) if is_tfr else (-2, -1)
```

Comment:
I am wondering about the EpochsTFR case, currently we only swap the third and first dimension for tfr evoked, but the same should be done for EpochsTFR, right? 
```suggestion
dimensions = [
    '3D' if len(series.shape) == 3 else '2D'
    if is_epo
]
    axes = (-3, -1) if (is_tfr or dimension == '3D') else (-2, -1)
```

Reply:
hmm. I'm inclined to say "I think it's correct as-is because the tests pass".  But then I remember that I messed around a lot with which axes to swap in the *old* API part of the test, to get it to match the new API results' shape.  So, either both this and the test are correct, or both this and the test are wrong. Let's try to reason through it:

- The reason we want to swap axes is to get the channels on the last dimension. We want that because it makes specifying the adjacency easier: we can pass an adjacency matrix for channels/vertices, and for other dims just use "immediate neighbor" adjacency. But *note that we technically don't have to do this*: in this tutorial the adjacency is computed for *the existing/unmodified shape of the EpochsTFR.data array* https://mne.tools/dev/auto_tutorials/stats-sensor-space/40_cluster_1samp_time_freq.html#define-adjacency-for-statistics.
- if we *are* going to swap axes, we should make sure that we also provide parameters like `max_step` for the non-channel dimensions. This is a bit tricky because we need to potentially do so for both time *and* frequency dimensions, but the function doesn't necessarily know which dimensions will be present.
- So assuming that we *do* swap axes, what matters is getting "channel" at the end. Does it matter whether we end up with "freq, time, chan" or "time, freq, chan"?  I think it *doesn't* matter, as long as we're careful about our bookkeeping for which order we have.
    - Epochs is (epo, chan, time), Evoked is (chan, time). In both cases swapping axis -1 (time) with axis -2 (chan) seems like the right solution (because we definitely want to keep "epo" as the first axis).
    - EpochsTFR is (epo, chan, freq, time). So swapping axis -1 (time) with axis -3 (chan) will give us (epo, time, freq, chan).
    - AverageTFR is (chan, freq, time). So swapping axis -1 (time) with axis -3 (chan) will give us (time, freq, chan).

So, I think what we're doing here is OK *as long as we keep track of dimension order for TFRs and act accordingly*.

Reply:
...wait, maybe I misunderstood your question.  Perhaps you think `is_tfr` is only for AverageTFR?  in fact `is_tfr` should be true for both EpochsTFR and AverageTFR, so the same dims (-3, -1) will get swapped for any kind of TFR object.

Reply:
I thought that is_tfr only handles AverageTFR, which might create issues. After checking the code again, I agree that it should be fine, EpochsTFR means is_tfr == True as well as is_epo == True.